src/worker/swirl/runner.rs

```diff
@@ -26,155 +23,153 @@ fn runnable<J: BackgroundJob>(ctx: &J::Context, payload: serde_json::Value) -> a
-        let (sender, receiver) = sync_channel(max_threads);
-        let mut pending_messages = 0;
-        loop {
-            let available_threads = max_threads - self.thread_pool.active_count();
-
-            let jobs_to_queue = if pending_messages == 0 {
-                // If we have no queued jobs talking to us, and there are no
-                // available threads, we still need to queue at least one job
-                // or we'll never receive a message
-                max(available_threads, 1)
-            } else {
-                available_threads
-            };
+    /// This returns a `RunningRunner` which can be used to wait for the workers to shutdown.
+    pub fn start(&self) -> anyhow::Result<RunningRunner> {
+        let handles = (0..self.num_workers)
+            .map(|i| {
+                let name = format!("background-worker-{i}");
+                info!(worker.name = %name, "Starting worker…");
 
-            for _ in 0..jobs_to_queue {
                 let worker = Worker {
                     connection_pool: self.connection_pool.clone(),
                     context: self.context.clone(),
                     job_registry: self.job_registry.clone(),
-                    sender: sender.clone(),
+                    shutdown_when_queue_empty: self.shutdown_when_queue_empty,
+                    poll_interval: self.poll_interval,
                 };
 
-                self.thread_pool.execute(move || worker.run_next_job())
-            }
+                std::thread::Builder::new()
+                    .name(name.clone())
+                    .spawn(move || {
+                        info_span!("worker", worker.name = %name).in_scope(|| worker.run())
+                    })
+            })
+            .collect::<Result<_, _>>()?;
 
-            pending_messages += jobs_to_queue;
-            match receiver.recv_timeout(self.job_start_timeout)? {
-                Event::Working => pending_messages -= 1,
-                Event::NoJobAvailable => return Ok(()),
-                Event::ErrorLoadingJob(e) => return Err(FetchError::FailedLoadingJob(e)),
-                Event::FailedToAcquireConnection(e) => {
-                    return Err(FetchError::NoDatabaseConnection(e));
-                }
-            }
-        }
+        Ok(RunningRunner { handles })
     }
 
     fn connection(&self) -> Result<DieselPooledConn, PoolError> {
         self.connection_pool.get()
     }
 
-    /// Waits for all running jobs to complete, and returns an error if any
-    /// failed
+    /// Check if any jobs in the queue have failed.
     ///
-    /// This function is intended for use in tests. If any jobs have failed, it
-    /// will return `swirl::JobsFailed` with the number of jobs that failed.
-    ///
-    /// If any other unexpected errors occurred, such as panicked worker threads
-    /// or an error loading the job count from the database, an opaque error
-    /// will be returned.
```

Comment:
While this was apparently wrong about the details before, do we want to retain the explicit documentation that an error gets returned when the check fails? (I guess it's _fairly_ obvious from the function signature, but I'm always conscious of the diminished brainpower in late night coding sessions. Or, in my case, early morning ones.)